raster-foundry/raster-foundry
# Prevent concurrent task unlocks from different backsplash instances

## Overview

This PR acquires an advisory lock using a new table called `last_unlocked` before automatically unlocking stuck tasks. The purpose of doing so is to prevent distinct unlocking runs from trying to unlock the same tasks, which causes the task status to toggle back and forth several times.

### Checklist


### Notes

Depends on #5502 because it also has a migration

Automated testing is really hard because of the rollback transactor -- I wasn't able to run two things in parallel in a way that preserved database state, since `ConnectionIO` doesn't have a `Parallel` instance, and running anything through `transact(xa)` causes the rollback to occur. I experimented with using the `Transactor.after` lens to build a new transactor just for this test, but I wasn't successful. However, the scaling test strategy did produce useful evidence.

This also includes a bunch of changelog updates because I believed I could make `chan` happy, but I did not make `chan` happy despite following the familiar playbook from yesterday (remove all the empty sections, remove extra returns after headings, check for bogus section heading names, add citations at the bottom)

## Testing Instructions

- assemble backsplash *on develop*
- apply this diff to prevent port forwarding from backsplash and to make the task lock expiration time really short

```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index 8b574422e..61d27db2f 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -176,10 +176,10 @@ services:
       - BACKSPLASH_ENABLE_MULTITIFF=false
       - AWS_REQUEST_PAYER=requester
       - DB_SCENE_SEARCH_BUFFER_PERCENTAGE=0.1
-      - DB_TASK_STATUS_EXPIRATION_SECONDS=3600
-    ports:
-      - "8080:8080"
-      - "9030:9030"
+      - DB_TASK_STATUS_EXPIRATION_SECONDS=10
+        # ports:
+            # - "8080:8080"
+            # - "9030:9030"
     volumes:
       - ./app-backend/:/opt/raster-foundry/app-backend/
       - ./data/:/opt/data/
```

- lock a task in your database: `update tasks set locked_by = 'default', locked_on = now() where id = '1f5f63dc-a42e-402e-9fd0-58e133435cb1';`
- bring up two copies of backsplash: `docker-compose up --scale=backsplash=2 backsplash`
- when the task unlock runs, observe that both backsplash instances warn you about the project id for the task you locked
- stop the servers
- checkout this branch and reassemble backsplash
- relock the task
- bring two servers up again
- when the task unlock runs, observe that only one of the backsplash instances warns you about the project id for the task you locked

Closes azavea/raster-foundry-platform#1120
